docs/design/paginators.md

```diff
@@ -68,71 +68,87 @@ structure FunctionConfiguration { ... }
+}
+
+inline fun<T, R> SdkAsyncIterable<T>.map(transform: (T) -> R): SdkAsyncIterable<R> = TODO("not-implemented")
+inline fun<T, R> SdkAsyncIterable<T>.mapNotNull(transform: (T) -> R?): SdkAsyncIterable<R> = TODO("not-implemented")
+inline fun<T, R> SdkAsyncIterable<Iterable<T>>.flatMap(transform: (T) -> Iterable<R>): SdkAsyncIterable<R> = TODO("not-implemented")
 
-inline fun<T, R> Paginator<T>.map(transform: (T) -> R): Paginator<R> = TODO("not-implemented")
-inline fun<T, R> Paginator<Iterable<T>>.flatMap(transform: (T) -> Iterable<R>): Paginator<R> = TODO("not-implemented")
-fun<T> Paginator<Iterable<T>>.flatten(): Paginator<T> = TODO("not-implemented")
-fun <T> Paginator<T?>.filterNotNull(): Paginator<T> = TODO("not-implemented")
-suspend fun<T> Paginator<T>.toList(): List<T> = TODO("not-implemented")
-suspend fun<T> Paginator<T>.toSet(): Set<T> = TODO("not-implemented")
+fun<T> SdkAsyncIterable<Iterable<T>>.flatten(): SdkAsyncIterable<T> = TODO("not-implemented")
+fun <T> SdkAsyncIterable<T?>.filterNotNull(): SdkAsyncIterable<T> = TODO("not-implemented")
+
+suspend fun<T> SdkAsyncIterable<T>.toList(): List<T> = TODO("not-implemented")
+suspend fun<T> SdkAsyncIterable<T>.toSet(): Set<T> = TODO("not-implemented")
+
+/**
+ * Terminal operator that collects the given iterable and calls [action] for each result
+ */
+inline fun <T> SdkAsyncIterable<T>.collect(crossinline action: suspend (T) -> Unit):Unit = TODO()
 ```
 
-Pagination by default will always generate a paginator _over the normal operation output type_. If `items` is specified then the generated paginator will be “specialized” to have an additional field that paginates only over the member targeted by `items`. An example of this is demonstrated below:
+Pagination by default will always generate an `SdkAsyncIterable` over the normal operation output type. 
+If `items` is specified then the generated paginator will be “specialized” to have an additional field that iterates 
+only over the member targeted by `items`. The targeted type will be flattened for the user automatically.
+
+An example of this is demonstrated below:
 
 
 ```kotlin
 class ListFunctionsPaginator(
     private val client: LambdaClient,
     private val initialRequest: ListFunctionsRequest
-): Paginator<ListFunctionsResponse>{
-    private var cursor: String? = null
-    private var isFirstPage: Boolean = true
+): SdkAsyncIterable<ListFunctionsResponse>, SdkAsyncIterator<ListFunctionsResponse> {
+   private var cursor: String? = null
+   private var isFirstPage: Boolean = true
 
-    override val hasMorePages: Boolean
-        get() = isFirstPage || (cursor?.isNotEmpty() ?: false)
+   override operator fun iterator(): SdkAsyncIterator<ListFunctionsResponse> = this
 
-    override suspend fun next(): ListFunctionsResponse {
-        if (!hasMorePages) throw NoSuchElementException("no pages remaining")
+   override operator fun hasNext(): Boolean
+           = isFirstPage || (cursor?.isNotEmpty() ?: false)
```

Comment:
Hmm that's a good point. It may be that `hasNext()` needs to be `suspend` as well because transformations of any kind off the original paginator will all suffer the same issue where we don't actually know if there is an item or not without getting the first page.

Reply:
The more I think about this, the more I wonder if there are actually _two_ types here:

A page-fetching iterator that knows _without suspending_ when more items are available:
```kotlin
interface AsyncResponseIterator<out T> {
    operator fun hasNext(): Boolean
    suspend operator fun next(): T
}
```

An item-fetching iterator that _may have to fetch one or more pages_ (i.e., suspend) to determine if there are more items:
```kotlin
interface AsyncItemsIterator<out T> {
    suspend operator fun hasNext(): Boolean
    suspend operator fun next(): T
}
```